build: Honour 'output_file' for cargo builds

joergroedel opened the pull request:
With newer versions of the Rust toolchain cargo started to rename the user-mode output binaries, which broke the build for the SVSM file-system image.

Fix that by honouring the 'output_file' parameter for cargo recipes and specifying an alternative build output name for userinit.

00xc commented:
I think changing the `userinit` binary name instead (as I did in #686) is less error prone, since there is no real reason for a package to produce an differently named binary, and even then one can use `--bin` instead of `--package` if it's really needed.

joergroedel commented:
> I think changing the `userinit` binary name instead (as I did in #686) is less error prone, since there is no real reason for a package to produce an differently named binary, and even then one can use `--bin` instead of `--package` if it's really needed (setting `"binary": true` would also fix this issue without changing the build tool).

You are right. I changed it as in PR #686 and removed the changes to the build recipes. But let the build script honour `output_file` is a good change for consistency as well, so I kept this change.

joergroedel commented:
> Is the build with `--release` now expected to work or is this a separate issue?
> 
> This is still failing for me:
> 
> ```
> ./build --release configs/qemu-target.json
> [...]
>    Compiling bootlib v0.1.0 (/home/osteffen/svsm/bootlib)
>    Compiling cpuarch v0.1.0 (/home/osteffen/svsm/cpuarch)
>    Compiling libtcgtpm v0.1.0 (/home/osteffen/svsm/libtcgtpm)
>     Finished `release` profile [optimized] target(s) in 19.35s
> Building stage2...
>    Compiling svsm v0.1.0 (/home/osteffen/svsm/kernel)
> error: linking with `ld` failed: exit status: 1
>   |
>   = note: LC_ALL="C" PATH="/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/osteffen/.cargo/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/osteffen/.local/bin:/home/osteffen/bin" VSLANG="1033" "ld" "/tmp/rustcW2YNoh/symbols.o" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/stage2-d01b7f5c5270a901.stage2.b735d2d3bd46ef15-cgu.0.rcgu.o" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/stage2-d01b7f5c5270a901.stage2.b735d2d3bd46ef15-cgu.1.rcgu.o" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/stage2-d01b7f5c5270a901.71w39tnc471615ss07oh6ff8w.rcgu.o" "--as-needed" "-Bstatic" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsvsm-60f30c60111a6578.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/liblog-c0be093504f6004c.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libuuid-45567d2ad2c502dd.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libintrusive_collections-bc78b24943396d1d.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libmemoffset-317ba3f39dd16912.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libigvm_defs-aa2b0ec16fea74f1.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libstatic_assertions-084f01c3d091ede0.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libopen_enum-b9729c04b2fc3ce7.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libpackit-de3d076cb96fe6f4.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libzerocopy-cdea71412bc605f6.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libbyteorder-ce14d8a13f79748b.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libelf-7328de20e95b308e.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsha2-c2babc296be5ca99.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libdigest-1401476d75411093.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libblock_buffer-3a439e35f2ac906b.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libaes_gcm-46b4566c7131412a.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libctr-c2eb14c86233e472.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libghash-b927bb42bcbbcc95.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libpolyval-73006a9aa1c16f6e.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libopaque_debug-dc91f4bdf1269f53.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libuniversal_hash-f3bd5f68f9c60163.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsubtle-567a3d1be0615487.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libaes-2ced0f712be57218.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libcfg_if-00eb171ddef518b0.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libcipher-a9f20e41eb7743c0.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libinout-4b353589a3f42edc.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libaead-91fd11dbc4557669.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libcrypto_common-8da0fa718484c08e.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libgeneric_array-4aebcdb35aaa5fc7.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libtypenum-d0e45a56182c6d6a.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsyscall-a4fbf40b8b3c23e6.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libbitflags-baf0f8598f3eae2b.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/librelease-bdf8a85db79ef658.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libcpuarch-b6aa5f6116bf8249.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libvstd-e46250c2184e2698.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libbuiltin-f8357583009bc8c9.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libbootlib-faa51ef04405d6a3.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libzerocopy-a1b5b4559f403a34.rlib" "/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-none/lib/liballoc-896004cd0d5424b6.rlib" "/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-none/lib/librustc_std_workspace_core-1e8b6240333f546f.rlib" "/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-none/lib/libcore-84e1ed6b861620f8.rlib" "/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-none/lib/libcompiler_builtins-bba1f7349cafb20e.rlib" "-Bdynamic" "--eh-frame-hdr" "-z" "noexecstack" "-o" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/stage2-d01b7f5c5270a901" "--gc-sections" "-pie" "-z" "relro" "-z" "now" "-O1" "--strip-debug" "-nostdlib" "--build-id=none" "-Tkernel/src/stage2.lds" "-no-pie"
>   = note: ld: /home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsvsm-60f30c60111a6578.rlib(svsm-60f30c60111a6578.svsm.fab90bb829174d2-cgu.10.rcgu.o): in function `ex_handler_page_fault_early':
>           svsm.fab90bb829174d2-cgu.10:(.text.ex_handler_page_fault_early+0x43): undefined reference to `early_exception_table_start'
>           ld: svsm.fab90bb829174d2-cgu.10:(.text.ex_handler_page_fault_early+0x60): undefined reference to `early_exception_table_end'
> 
>   = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
>   = note: use the `-l` flag to specify native libraries to link
>   = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-link-lib)
> 
> error: could not compile `svsm` (bin "stage2") due to 1 previous error
> Error: configs/qemu-target.json: Command '['cargo', 'build', '--bin', 'stage2', '--manifest-path', 'kernel/Cargo.toml', '--release', '--target', 'x86_64-unknown-none']' returned non-zero exit status 101.
> ```

Hmm, that is a separate issue, but probably related to the compiler update as well.

00xc commented:
> > I think changing the `userinit` binary name instead (as I did in #686) is less error prone, since there is no real reason for a package to produce an differently named binary, and even then one can use `--bin` instead of `--package` if it's really needed (setting `"binary": true` would also fix this issue without changing the build tool).
> 
> You are right. I changed it as in PR #686 and removed the changes to the build recipes. But let the build script honour `output_file` is a good change for consistency as well, so I kept this change.

This should only be honored for make targets. For cargo I don't see how the resulting binary would be named differently than the binary or the package.

00xc commented:
> Is the build with `--release` now expected to work or is this a separate issue?
> 
> This is still failing for me:
> 
> ```
> ./build --release configs/qemu-target.json
> [...]
>    Compiling bootlib v0.1.0 (/home/osteffen/svsm/bootlib)
>    Compiling cpuarch v0.1.0 (/home/osteffen/svsm/cpuarch)
>    Compiling libtcgtpm v0.1.0 (/home/osteffen/svsm/libtcgtpm)
>     Finished `release` profile [optimized] target(s) in 19.35s
> Building stage2...
>    Compiling svsm v0.1.0 (/home/osteffen/svsm/kernel)
> error: linking with `ld` failed: exit status: 1
>   |
>   = note: LC_ALL="C" PATH="/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/osteffen/.cargo/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/osteffen/.local/bin:/home/osteffen/bin" VSLANG="1033" "ld" "/tmp/rustcW2YNoh/symbols.o" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/stage2-d01b7f5c5270a901.stage2.b735d2d3bd46ef15-cgu.0.rcgu.o" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/stage2-d01b7f5c5270a901.stage2.b735d2d3bd46ef15-cgu.1.rcgu.o" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/stage2-d01b7f5c5270a901.71w39tnc471615ss07oh6ff8w.rcgu.o" "--as-needed" "-Bstatic" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsvsm-60f30c60111a6578.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/liblog-c0be093504f6004c.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libuuid-45567d2ad2c502dd.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libintrusive_collections-bc78b24943396d1d.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libmemoffset-317ba3f39dd16912.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libigvm_defs-aa2b0ec16fea74f1.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libstatic_assertions-084f01c3d091ede0.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libopen_enum-b9729c04b2fc3ce7.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libpackit-de3d076cb96fe6f4.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libzerocopy-cdea71412bc605f6.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libbyteorder-ce14d8a13f79748b.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libelf-7328de20e95b308e.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsha2-c2babc296be5ca99.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libdigest-1401476d75411093.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libblock_buffer-3a439e35f2ac906b.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libaes_gcm-46b4566c7131412a.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libctr-c2eb14c86233e472.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libghash-b927bb42bcbbcc95.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libpolyval-73006a9aa1c16f6e.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libopaque_debug-dc91f4bdf1269f53.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libuniversal_hash-f3bd5f68f9c60163.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsubtle-567a3d1be0615487.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libaes-2ced0f712be57218.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libcfg_if-00eb171ddef518b0.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libcipher-a9f20e41eb7743c0.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libinout-4b353589a3f42edc.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libaead-91fd11dbc4557669.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libcrypto_common-8da0fa718484c08e.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libgeneric_array-4aebcdb35aaa5fc7.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libtypenum-d0e45a56182c6d6a.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsyscall-a4fbf40b8b3c23e6.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libbitflags-baf0f8598f3eae2b.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/librelease-bdf8a85db79ef658.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libcpuarch-b6aa5f6116bf8249.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libvstd-e46250c2184e2698.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libbuiltin-f8357583009bc8c9.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libbootlib-faa51ef04405d6a3.rlib" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libzerocopy-a1b5b4559f403a34.rlib" "/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-none/lib/liballoc-896004cd0d5424b6.rlib" "/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-none/lib/librustc_std_workspace_core-1e8b6240333f546f.rlib" "/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-none/lib/libcore-84e1ed6b861620f8.rlib" "/home/osteffen/.rustup/toolchains/1.84.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-none/lib/libcompiler_builtins-bba1f7349cafb20e.rlib" "-Bdynamic" "--eh-frame-hdr" "-z" "noexecstack" "-o" "/home/osteffen/svsm/target/x86_64-unknown-none/release/deps/stage2-d01b7f5c5270a901" "--gc-sections" "-pie" "-z" "relro" "-z" "now" "-O1" "--strip-debug" "-nostdlib" "--build-id=none" "-Tkernel/src/stage2.lds" "-no-pie"
>   = note: ld: /home/osteffen/svsm/target/x86_64-unknown-none/release/deps/libsvsm-60f30c60111a6578.rlib(svsm-60f30c60111a6578.svsm.fab90bb829174d2-cgu.10.rcgu.o): in function `ex_handler_page_fault_early':
>           svsm.fab90bb829174d2-cgu.10:(.text.ex_handler_page_fault_early+0x43): undefined reference to `early_exception_table_start'
>           ld: svsm.fab90bb829174d2-cgu.10:(.text.ex_handler_page_fault_early+0x60): undefined reference to `early_exception_table_end'
> 
>   = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
>   = note: use the `-l` flag to specify native libraries to link
>   = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-link-lib)
> 
> error: could not compile `svsm` (bin "stage2") due to 1 previous error
> Error: configs/qemu-target.json: Command '['cargo', 'build', '--bin', 'stage2', '--manifest-path', 'kernel/Cargo.toml', '--release', '--target', 'x86_64-unknown-none']' returned non-zero exit status 101.
> ```

This works for me on 1.84.1 using #686

joergroedel commented:
> This should only be honored for make targets. For cargo I don't see how the resulting binary would be named differently than the binary or the package.

There is still a difference between the package name and the filename of the resulting binary. Currently the script assumes that both are the same, but as we have seen they can be different when configured in `Cargo.toml`.

joergroedel commented:
> > This works for me on 1.84.1 using #686
> 
> Yeah, #686 works for me: `cargo xbuild --release configs/qemu-target.json`.
> 
> It also works to use the `build.py` from this PR on top of #686.

Hmm, interesting. Any clue what specific change in this PR fixes the issue?

00xc commented:
> > > This works for me on 1.84.1 using #686
> > 
> > 
> > Yeah, #686 works for me: `cargo xbuild --release configs/qemu-target.json`.
> > It also works to use the `build.py` from this PR on top of #686.
> 
> Hmm, interesting. Any clue what specific change in this PR fixes the issue?

My guess would be something in the first commit. I think in Oliver's error cargo is trying to use the userspace target triple.

joergroedel commented:
> My guess would be something in the first commit. I think in Oliver's error cargo is trying to use the userspace target triple.

Okay, anyhow, this is a separate issue and we should track it elsewhere. Is there any objection against the fix in this PR? Given what introduced the issue I think it is the right one.
